docs/csharp/pattern-matching.md

```diff
@@ -2,31 +2,192 @@
+types of shapes to this library, let's change the existing method to use
+the `switch` pattern match expression:
+
+[!code-csharp[ComputeWithSwitchExpression](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#06_ComputeWithSwitchExpression "Compute with switch expression")]
+
+The pattern matching `switch` statement uses familiar syntax to developers
+that have used the traditional `switch` statement. Each `case` is evaluated
+and the code beneath the condition that matches the input variable is
+executed. Code execution cannot "fall through" from one `case` expression
+to the next. Each block following a `case` must end with a `break`, `return`
+or `goto`.
+
+There are important new rules governing the `switch` statement. The restrictions
+on the type of the variable in the `switch` expression have been removed.
+Any type, such as `object` in this example, may be used. The case expressions
+are no longer limited to constant values. Removing that limitation means
+that reordering switch sections may change a program's behavior.
+
+When limited to constant values, no more than one switch
+label could match the value of the switch expression. Combine that with the 
+rule that every switch section must not fall through to the next section, and 
+it followed that the
+switch sections could be rearranged in any order without affecting behavior.
+Now, with more generalized switch expressions, the order of each section
+matters. The switch expressions are evaluated in textual order. Execution
+transfers to the first switch label that matches the switch expression.  
+Note that the `default` case will only be executed if no other
+case labels match. The `default` case is evaluated last, regardless
+of its textual order.
+
+## `when` clauses in switch expressions
+
+You can make special cases for those shapes that have 0 area by using
+a `when` clause on the switch label. A square with side length of 0, or
+a circle with a radius of 0 has a 0 area. You specify that condition
+using a `when` clause on the switch label:  
+
+[!code-csharp[ComputeDegenerateShapes](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#07_ComputeDegenerateShapes "Compute shapes with 0 area")]
+
+This change demonstrates a few important points on the new syntax. First,
+multiple switch labels can be applied to one switch section. The statement
+block is executed when any of those labels is true. In this instance,
+if the switch expression is either a circle or a square with 0 area, the
+method returns the constant 0.
+
+This example introduces two different variables in the two switch labels
+for the first switch block. Notice that the statements in this switch block
+do not use either the variables `c` (for the circle), or `s` (for the square).
+Neither of those variables is definitely assigned in this switch block.
+Clearly one has been assigned, but static analysis cannot know which one. For that reason,
```

Comment:
> Clearly one has been assigned, but static analysis cannot know which one.

Suggested change:

> If either of these cases match, clearly one of the variables has been assigned.  However, it is impossible to tell *which* has been assigned at compile-time, because either case could match at runtime.  For that reason, ...